rust/src/alive_test/alive_test.rs

```diff
@@ -0,0 +1,324 @@
+// SPDX-FileCopyrightText: 2024 Greenbone AG
+//
+// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception
+
+use crate::alive_test::AliveTestError;
+use crate::models::{AliveTestMethods, Host};
+
+use futures::StreamExt;
+use pnet::packet::ip::IpNextHeaderProtocols;
+use std::time::Duration;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::time::sleep;
+
+use std::{
+    net::{IpAddr, Ipv4Addr, SocketAddr},
+    str::FromStr,
+};
+
+use pcap::{Active, Capture, Inactive, PacketCodec, PacketStream};
+use pnet::packet::{
+    self,
+    icmp::{IcmpCode, IcmpTypes, MutableIcmpPacket, *},
+    ip::IpNextHeaderProtocol,
+    ipv4::{checksum, Ipv4Packet, MutableIpv4Packet},
+};
+
+use socket2::{Domain, Protocol, Socket};
+
+const IPPROTO_RAW: i32 = 255;
+const DEFAULT_TIMEOUT_MS: u64 = 5000;
+const ICMP_LENGTH: usize = 8;
+const IP_LENGTH: usize = 20;
+const HEADER_LENGTH: u8 = 5;
+const DEFAULT_TTL: u8 = 255;
+const MIN_ALLOWED_PACKET_LEN: usize = 16;
+enum AliveTestCtl {
+    Stop,
+    // (IP and successful detection method)
+    Alive {
+        ip: String,
+        detection_method: AliveTestMethods,
+    },
+}
+fn make_mut_icmp_packet(buf: &mut Vec<u8>) -> Result<MutableIcmpPacket, AliveTestError> {
+    MutableIcmpPacket::new(buf).ok_or_else(|| AliveTestError::CreateIcmpPacket)
+}
+
+fn new_raw_socket() -> Result<Socket, AliveTestError> {
+    Socket::new_raw(
+        Domain::IPV4,
+        socket2::Type::RAW,
+        Some(Protocol::from(IPPROTO_RAW)),
+    )
+    .map_err(|e| AliveTestError::NoSocket(e.to_string()))
+}
+
+fn forge_icmp(dst: IpAddr) -> Result<Vec<u8>, AliveTestError> {
+    if dst.is_ipv6() {
+        return Err(AliveTestError::InvalidDestinationAddr);
+    }
+
+    // Create an icmp packet from a buffer and modify it.
+    let mut buf = vec![0; ICMP_LENGTH];
+    let mut icmp_pkt = make_mut_icmp_packet(&mut buf)?;
+    icmp_pkt.set_icmp_type(IcmpTypes::EchoRequest);
+    icmp_pkt.set_icmp_code(IcmpCode::new(0u8));
+
+    // Require an unmutable ICMP packet for checksum calculation.
+    // We create an unmutable from the buffer for this purpose
+    let icmp_aux = IcmpPacket::new(&buf).ok_or_else(|| AliveTestError::CreateIcmpPacket)?;
+    let chksum = pnet::packet::icmp::checksum(&icmp_aux);
+
+    // Because the buffer of original mutable icmp packet is borrowed,
+    // create a new mutable icmp packet to set the checksum in the original buffer.
+    let mut icmp_pkt = make_mut_icmp_packet(&mut buf)?;
+    icmp_pkt.set_checksum(chksum);
```

Comment:
You can save all of the auxilliary packets via
```suggestion rust
    // Create an icmp packet from a buffer and modify it.
    let mut buf = vec![0; ICMP_LENGTH];
    let mut icmp_pkt = make_mut_icmp_packet(&mut buf)?;
    icmp_pkt.set_icmp_type(IcmpTypes::EchoRequest);
    icmp_pkt.set_icmp_code(IcmpCode::new(0u8));
    icmp_pkt.set_checksum(icmp::checksum(&icmp_pkt.to_immutable()));
```

(note: I imported `pnet::packet::icmp`, but we can also qualify the path if you prefer)